client/app/components/QuerySelector.jsx

```diff
@@ -0,0 +1,149 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import { debounce, find } from 'lodash';
+import Input from 'antd/lib/input';
+import { Query } from '@/services/query';
+import { toastr } from '@/services/ng';
+import { QueryTagsControl } from '@/components/tags-control/TagsControl';
+
+const SEARCH_DEBOUNCE_DURATION = 200;
+
+function search(term) {
+  // get recent
+  if (!term) {
+    return Query.recent().$promise
+      .then((results) => {
+        const filteredResults = results.filter(item => !item.is_draft); // filter out draft
+        return Promise.resolve(filteredResults);
+      });
+  }
+
+  // search by query
+  return Query.query({ q: term }).$promise
+    .then(({ results }) => Promise.resolve(results));
+}
+
+export function QuerySelector(props) {
+  const [searchTerm, setSearchTerm] = useState();
+  const [searching, setSearching] = useState();
+  const [searchResults, setSearchResults] = useState([]);
+  const [selectedQuery, setSelectedQuery] = useState();
+
+  let isStaleSearch = false;
+  const debouncedSearch = debounce(_search, SEARCH_DEBOUNCE_DURATION);
+  const placeholder = 'Search a query by name';
+  const clearIcon = <i className="fa fa-times" onClick={() => setSelectedQuery(null)} />;
+  const spinIcon = <i className={cx('fa fa-spinner fa-pulse', { hidden: !searching })} />;
+
+  // set selected from prop
+  useEffect(() => {
+    if (props.selectedQuery) {
+      setSelectedQuery(props.selectedQuery);
+    }
+  }, [props.selectedQuery]);
+
+  // on search term changed, debounced
+  useEffect(() => {
+    // clear results, no search
+    if (searchTerm === null) {
+      setSearchResults(null);
+      return () => {};
+    }
+
+    // search
+    debouncedSearch(searchTerm);
+    return () => {
+      debouncedSearch.cancel();
+      isStaleSearch = true;
+    };
+  }, [searchTerm]);
+
+  // on query selected/cleared
+  useEffect(() => {
+    setSearchTerm(selectedQuery ? null : ''); // empty string forces recent fetch
+    props.onChange(selectedQuery);
+  }, [selectedQuery]);
+
+  function _search(term) {
+    setSearching(true);
+    search(term)
+      .then(rejectStale)
+      .then(setSearchResults)
+      .finally(() => {
+        setSearching(false);
+      });
+  }
+
+  function rejectStale(results) {
+    return isStaleSearch
+      ? Promise.reject(new Error('stale'))
+      : Promise.resolve(results);
+  }
+
+  function selectQuery(queryId) {
+    const query = find(searchResults, { id: queryId });
+    if (!query) { // shouldn't happen
+      toastr.error('Something went wrong.. Couldn\'t select query');
```

Comment:
Seems you didn't decide what to use - single dot or three dots, so used something between :smile:

Reply:
Fixed!